Add registry pattern to eth-abi

davesque opened the pull request:
### What was wrong?

Need to add registry pattern to eth-abi library.

### How was it fixed?

Not done yet.  Work in progress.

#### Cute Animal Picture

![Cute animal picture](http://justsomething.co/wp-content/uploads/2014/01/animal-photobombs-14-ft.jpg)

pipermerriam commented on eth_abi/registry.py:
```diff
@@ -0,0 +1,146 @@
+            return
+
+        # Is string lookup
+        if lookup in self._encoder_registry:
```
We've generally tried to write our conditionals to be explicit in the following manner.

```python
if callable(thing):
    # do something
elif isinstance(thing, str):
    # do something
else:
    raise TypeError("<good error message indicating type not supported, what type was given, what types are allowed")
```

It ensures we don't hit weird cases where someone uses `bytes` or an `int` or something and our registry ends up with weird values.

davesque replied:
Sounds good.  Just FYI, part of my deciding to do things that way had to do with similar duck-typing I saw here:
https://github.com/ethereum/eth-abi/blob/master/eth_abi/abi.py#L26-L29

Should we consider being more explicit in that case as well?

pipermerriam replied:
Yes, that would be an improvement.  Adding type hints here (and mypy typing would also help but that should probably be done independently).

pipermerriam commented on eth_abi/registry.py:
```diff
@@ -0,0 +1,146 @@
+
+class ABIRegistry:
+    def __init__(self):
+        self._encoder_registry = {}
```
*naming nitpick*

Since the class is the registry, this seems more appropriately named `_encoders` since these are the registered encoders.

davesque replied:
Good suggestion.  The naming there has to do with the fact that I originally had a single dict just called `_registry`.  I'll change it.

pipermerriam commented on eth_abi/registry.py:
```diff
@@ -0,0 +1,146 @@
+
+def ensure_changed(ExcClass, error_msg):
+    def decorator(old_method):
+        def new_method(self, *args, **kwargs):
```
*naming convention nitpick*

We've got a somewhat consistent convention for decorators to use the names `fn`, `inner` and `outer`

```python
def simple_dec(fn):
    def inner(...):
         return fn(...)
    return inner

def args_dec(arg):
    def outer(fn):
         def inner(...):
             return fn(...)
        return inner
    return outer
```

I don't know if we have many in place that are method decorators, so no established convention for naming of the method fn.  I'd suggest something like `meth`, `method` or something else better I haven't though of.

davesque replied:
Sure, I can use that convention.  The naming convention I've been using is geared toward being as descriptive as possible but still reasonably concise which is hopefully helpful to people who are less strong with decorators.  I've seen names like `outer`, `inner` but I generally don't find them to be very meaningful.  But I'm happy to go with that if it's already an established pattern.

pipermerriam replied:
I can get onboard with that.  +1 (no need to go update the other cases, we can just migrate our convention over time when that code gets touched).

pipermerriam commented on eth_abi/registry.py:
```diff
@@ -0,0 +1,146 @@
+        self._items = [i for i in self._items if i.predicate is not predicate]
+
+    @ensure_changed(KeyError, 'No matching items found')
+    def remove_matching(self, *args, **kwargs):
```
This doesn't appear to be used.

davesque replied:
Yep, things aren't done yet so I just have that in there in case I want to unregister callbacks in a different way.

pipermerriam commented on eth_abi/registry.py:
```diff
@@ -0,0 +1,146 @@
+    def remove_matching(self, *args, **kwargs):
+        self._items = [i for i in self._items if not i.predicate(*args, **kwargs)]
+
+    def __contains__(self, item):
```
Maybe I missed it but I don't see any `thing in predicate_list` code here (but I could have easily missed it).

davesque replied:
You're right.  I need to take that out.  I had previously been doing explicit checks for membership somewhere but then switched to catching key errors instead.

pipermerriam reviewed (commented):
Generally good.  

I question the need for the `PredicateList` class.  Can we get away with only having the `_encoder_registry` and `_decoder_registry`?  It seems like we should be able to key on the `lookup`.

- custom functions/callables can just be identified by `id`
- string lookups can be either pre-processed into a callable which is unique based on the original string **or** we can just store the string and deal with the matching at some later stage when we try to lookup an encoder/decode.

Thoughts?

carver commented on eth_abi/registry.py:
```diff
@@ -0,0 +1,94 @@
+            self._encoders.add(lookup, encoder, label)
+            return
+
+        if isinstance(lookup, str):
```
This is a little more awkward with the predicate approach, but probably still worth doing: 

We talked about checking the `str` against existing `str` predicates at register-time. Unfortunately, I can't think of a better way to check the predicates than at encode time.

davesque replied:
Need some clarification on this.  Are you wondering how we'll detect conflicts at register time if everything is a predicate?

carver replied:
Actually, I missed that this was adding the string `lookup` as a label. So resolving the other comment will address this issue.

carver replied:
I was just wanting to make sure that registering encoders twice with the same string would fail.

carver commented on eth_abi/registry.py:
```diff
@@ -0,0 +1,94 @@
+        if label is not None:
+            self._labeled_predicates[label] = predicate
+
+        self._values[predicate] = value
```
I think this should raise if the label or predicate already exist.

davesque replied:
Yep, thanks for the heads up.

carver reviewed (commented):
Direction seems good to me. Just a couple safety checks to consider.

pipermerriam commented on eth_abi/registry.py:
```diff
@@ -0,0 +1,100 @@
+            return
+
+        if isinstance(lookup, str):
+            def predicate(base, sub, arrlist):
```
Extracting this to the module level and currying it should make this a bit cleaner.

```python
from cytoolz import curry

@curry
def static_predicate(lookup, base, sub, arrlist):
    return collapse_type(base, sub, arrlist) == lookup

... # then using it later.
self.encoders.add(static_predicate(lookup), encoder, lookup)
```

davesque replied:
I guess that means we'd need to add cytoolz under `install_requires` in `setup.py`.  Are there any problems that might come up as a result of both eth-abi and eth-utils depending on cytoolz (e.g. different version specs for cytoolz package might become out of sync, etc.)?  I guess I've worked on other projects where this might have come up but never thought of this before.

pipermerriam replied:
`cytoolz` is pretty standard across our codebases as well as being quite stable.  Maybe worth doing a quick check that the version ranges intersect with whatever is specified in `eth-utils`

davesque commented on eth_abi/registry.py:
```diff
@@ -0,0 +1,126 @@
+                raise NoEntryFound('No entry found for given predicate')
+            return
+
+        if isinstance(lookup_or_label, str):
```
@pipermerriam Should I maybe be using `eth_utils.is_text` here?

pipermerriam replied:
I don't think it's needed.  Now that we're in python3 and differentiating between text/bytes doesn't require a check for what python version we're in, I think we can move away from `is_text` and `is_bytes` and potentially even `is_string`.  For lower level libs like this one, reducing the call-stack overhead should yield some basic albeit minor performance gains.

pipermerriam commented on eth_abi/registry.py:
```diff
@@ -0,0 +1,131 @@
+from .utils.parsing import collapse_type
+
+
+class ABIRegistryError(Exception):
```
We've typically located these in `lib_name.exceptions` by convention.

davesque replied:
I did notice that and planned to move the exceptions later.  Was just keeping everything in one file while I work on it.  But I'll go ahead and move it there now.

pipermerriam commented on eth_abi/registry.py:
```diff
@@ -0,0 +1,131 @@
+
+        self._values[predicate] = value
+
+    def find(self, *args, **kwargs):
```
Is the `*args, **kwargs` here because you aren't yet sure what the arguments will be, or to make this *generic*?  I think I'm ok with us making this implementation specific to whatever we *know* the arguments to our `predicate` functions are going to be so that we can eliminate one axis of validation/errors that this is able to produce by forcing things to fail at the `find(..)` invocation rather than somewhere deeper in the `predicate(...)` invocation.

davesque replied:
It was to make it generic but we can make it specific to its intended use.  I'll change it.

pipermerriam commented on eth_abi/registry.py:
```diff
@@ -0,0 +1,131 @@
+    def find(self, *args, **kwargs):
+        results = tuple(
+            value for predicate, value in self._values.items()
+            if predicate(*args, **kwargs)
```
*nitpick*

I think that this tuple comprehension is equivalent to the [`toolz.keyfilter`](http://toolz.readthedocs.io/en/latest/api.html#toolz.dicttoolz.keyfilter).  Always questionable whether using such functions is an improvement as they are often a readability trade off.  Merely pointing it out so you know the tool exists.

pipermerriam commented on eth_abi/registry.py:
```diff
@@ -0,0 +1,131 @@
+    def _label_for_predicate(self, predicate):
+        for key, value in self._labeled_predicates.items():
+            if value is predicate:
+                return key
```
Does this data structure contain an *implicit* guarantee that this loop either matches 1 or 0 values and cannot match multiple predicates?  If so, it might be worth adding a comment here to ensure future readers of this code know that case is not something that was *overlooked*.

davesque replied:
It does give that guarantee since both keys and values in that dict must be unique (or an exception is raised).  I'll add a comment to that effect.

pipermerriam commented on tests/registry/test_predicate_mapping.py:
```diff
@@ -0,0 +1,114 @@
+    return PredicateMapping()
+
+
+def gt(x, y):
```
Are you familiar with the [`operator`](https://docs.python.org/3/library/operator.html#operator.gt) module?

davesque replied:
Yes I am.  I can use it here if you think that makes things more readable.

davesque replied:
Actually that stuff is going to get taken out since I'm assuming only one argument for predicates.  I had a discussion with @carver and he'd like to use type strings as the canonical form for types.  So that means predicates would only ever receive one string as an argument.  They can parse that string as much as they like in order to figure out if they're interested in it.  We're planning on memoizing the type lookup so the added performance cost of the repeated parsing shouldn't be a problem.  I'm fine with going in that direction.  Thoughts?

pipermerriam commented on tests/registry/test_predicate_mapping.py:
```diff
@@ -0,0 +1,114 @@
+
+    assert mapping.find(2, 1) == 'greater than'
+    assert mapping.find(2, 2) == 'less than or equal'
+    assert mapping.find(2, 3) == 'less than or equal'
```
This isn't a big deal at all since these are so simple, but traditionally these would be done with the `pytest.mark.parametrize` decorator.  Just making sure you're aware of it's existence/usage.

davesque replied:
I did notice you guys using that decorator and thought I might use it in a case where there are many repetitions of some assertion.  I can use it there it you'd like.

davesque replied:
Actually I'm remembering now that the reason I did that was I thought it would be nice to test if the `find` method works if more than one predicate is present in the mapping.

pipermerriam commented on tests/registry/test_predicate_mapping.py:
```diff
@@ -0,0 +1,114 @@
+    with pytest.raises(ValueError) as e:
+        mapping.add(lte, 'bar', 'my_label')
+
+    e.match('Predicate with label')
```
Are you familiar with the `match` kwarg that `pytest.raises` accepts?

https://docs.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions

It should allow this to be written more succinctly as:

```python
with pytest.raises(ValueError, match="Predicate with label * already exists"):
    mapping.add(lte, 'bar', 'my_label')
```

davesque replied:
I wasn't familiar with that.  I haven't used pytest extensively so I'm happy to hear about any cool bells and whistles.

pipermerriam commented on eth_abi/registry.py:
```diff
@@ -0,0 +1,127 @@
+        try:
+            del self._values[predicate]
+        except KeyError:
+            raise KeyError('Predicate not found')
```
Seems like including the `str` representation of the `predicate` in the error here would be beneficial under the hope that our predicate functions are at least sometimes named in helpful ways.

davesque replied:
Just ignore the comments I made in reply to this.  I hadn't read your comment thoroughly enough.

pipermerriam replied:
You are correct that I mean `str(func_or_lambda)`.  I think this would be well complimented by changing the `equals` function as follows.

```python
def equals(value):
    def inner(other):
        return other == value
    inner.__name__ = "is_equal_to_{0}".format(value)
    return inner
```

This way our auto-generated functions will have helpful string representations.

One thing that isn't ideal about this is that `equals('arst') != equals('arst')` since the generated functions won't be equivalent.  I think we can fix this with some class shenanigans.

```python
class Equals:
    def __init__(self, value):
        self.value = value
    def __call__(self, other):
        return self.value == other
    def __eq__(self, other):
        return isinstance(other, Equals) and self.value == other.value
```

There are probably a few more dunder methods that need to be implemented like `__hash__`.

davesque replied:
It is possible that we can forego all of that by supplying labels for the default predicates?  My thinking was that, if people wanted to unregister default behavior, they would do so via one of the informatively chosen labels that we include.  I was also thinking of having a method on the registry that will provide a nice summary of what predicates are registered and any labels they might have.

If not, I think the idea of a callable `Equals` class is kinda cool anyway and I'd be happy to go with it.  It just seems like we could potentially avoid some of the edge cases that might come out of that approach by deferring to the labeling functionality.

Thoughts?

pipermerriam replied:
>  I was also thinking of having a method on the registry that will provide a nice summary of what predicates are registered and any labels they might have.

Implementing this with `__str__` seems appropriate.

davesque replied:
Went ahead and just took your suggestion here of having the `Equals` class.

pipermerriam commented on eth_abi/registry.py:
```diff
@@ -0,0 +1,127 @@
+        elif isinstance(predicate_or_label, str):
+            self.remove_by_label(predicate_or_label)
+        else:
+            raise TypeError('Key to be removed must be a callable or a string')
```
This is another place where including the type that was received in the error message makes debugging this failure path easier later.  Something like `"...: Got {0}".format(type(predicate_or_label))`

pipermerriam commented on eth_abi/utils/grammar.py:
```diff
@@ -0,0 +1,270 @@
+
+
+class TupleType(ABIType):
+    __slots__ = ('components',)
```
omg.  How did I not know about this!

http://book.pythontips.com/en/latest/__slots__magic.html

Looks like a 40-50% reduction in memory footprint!

davesque replied:
Yeah it seems like `__slots__` is ideal for small "immutable" container classes that only have a few attributes and many instances.  Fast constant time attribute lookups and space savings ftw.

davesque replied:
Although Python dicts in 3.6 share keys and are compact.  So I think it's less and less of a problem to just use normal classes with `__dict__` attrs.  Using `__slots__` is still a neat trick to have in your bag though.

pipermerriam replied:
Also, it's looking like our `mypy` based type checking is going to be incompatible with `__slots__`.  See https://github.com/ethereum/py-evm/issues/520

davesque replied:
Is that still the consensus?  It seemed that @carver had found a way to make it work.

pipermerriam replied:
Fine to leave in for now until we implement type hinting at which point we can see if we can use @carver 's solution.

pipermerriam reviewed (commented):
Looks like this is coming along.  It wasn't entirely clear to me reading the code what all of the various classes are for.  Wondering if you could take a few minutes to write up some basic docstrings where you think it makes sense.

pipermerriam commented on eth_abi/registry.py:
```diff
@@ -0,0 +1,254 @@
+
+        raise TypeError('Lookup/label must be a callable or a value of type `str`')
+
+    @classmethod
```
looks like this could be `staticmethod`

pipermerriam commented on eth_abi/registry.py:
```diff
@@ -0,0 +1,254 @@
+
+        raise TypeError('Lookup must be a callable or a value of type `str`')
+
+    @classmethod
```
looks like this could be `staticmethod`

pipermerriam commented on eth_abi/registry.py:
```diff
@@ -0,0 +1,254 @@
+        self._encoders = PredicateMapping()
+        self._decoders = PredicateMapping()
+
+    @classmethod
```
looks like this could be `staticmethod`

pipermerriam commented on eth_abi/utils/grammar.py:
```diff
@@ -0,0 +1,325 @@
+
+        return tuple(visited_children)
+
+    @functools.lru_cache(maxsize=None)
```
Any reason for the `None`?  Seems like it should be fine to put an arbitrary upper bound on this (1024?) so that we **know** it can't grow indefinitely.

davesque replied:
I just figured that, in most systems using eth-abi, there would probably only be on the order of a few hundred type strings at most which would ever be parsed.  Also, with `maxsize=None`, `functools.lru_cache` foregoes a lot of machinery that it would normally use to properly evict things from the cache.  Since I figured the cache would probably never grow too large, I thought it would be better to opt for the speedier version which doesn't limit size.

pipermerriam commented on tests/property/test_predicates.py:
```diff
@@ -46,6 +47,7 @@ def test_equals_has_expected_repr():
 
 @settings(max_examples=200)
 @given(type_strs)
+@example('int0')
```
Is `int0` a valid type ?  in fact is anything with a sub of `0` valid?

davesque replied:
No, it's not.  But it's parsable.  I'm pretty sure it could be parsed by the regexes we were using as well.  I guess I could update the grammar so that "0" is not a parsable sub component.  But it seemed like we were already drawing a pretty clear line between parsability and validity.

davesque committed:
Remove uses of deprecated hypothesis functionality

davesque committed:
Add parsimonious parser

davesque committed:
Add registry and `from_type_str` API

davesque committed:
Swap in registry code and delete legacy code

davesque committed:
Fix bug in BaseRealDecoder

There should never be any case in which the data for real values is not
32 bytes long.

davesque committed:
Minimum list size is 1 according to ABI spec

davesque committed:
Support parsing of 0-tuples

davesque committed:
Add some explicit test cases for 0-tuples
